panther_hardware_interfaces/src/panther_imu.cpp

```diff
@@ -0,0 +1,507 @@
+    }
+  }
+  return true;
+}
+
+void PantherImuSensor::ConfigureCompassParams()
+{
+  if (AreParamsDefined(
+        {"cc_mag_field", "cc_offset0", "cc_offset1", "cc_offset2", "cc_gain0", "cc_gain1",
+         "cc_gain2", "cc_t0", "cc_t1", "cc_t2", "cc_t3", "cc_t4", "cc_t5"})) {
+    ReadCompassParams();
+
+    spatial_->setCompassCorrectionParameters(
+      params_.cc_mag_field, params_.cc_offset0, params_.cc_offset1, params_.cc_offset2,
+      params_.cc_gain0, params_.cc_gain1, params_.cc_gain2, params_.cc_t0, params_.cc_t1,
+      params_.cc_t2, params_.cc_t3, params_.cc_t4, params_.cc_t5);
+  } else {
+    RCLCPP_INFO(logger_, "No compass correction params found. Skipping...");
+  }
+}
+
+void PantherImuSensor::ConfigureHeating()
+{
+  if (IsParamDefined("heating_enabled")) {
+    params_.heating_enabled =
+      hardware_interface::stod(info_.hardware_parameters["heating_enabled"]);
+    spatial_->setHeatingEnabled(params_.heating_enabled);
+  } else {
+    RCLCPP_INFO(logger_, "No heating enabled found. Skipping...");
+  }
+}
+
+void PantherImuSensor::ConfigureMadgwickFilter()
+{
+  ReadMadgwickFilterParams();
+
+  filter_.setWorldFrame(world_frame_);
+  filter_.setAlgorithmGain(params_.gain);
+  filter_.setDriftBiasGain(params_.zeta);
+}
+
+geometry_msgs::msg::Vector3 PantherImuSensor::ParseMagnitude(const double magnetic_field[3])
+{
+  geometry_msgs::msg::Vector3 mag_fld;
+
+  if (magnetic_field[0] == KImuMagneticFieldUnknownValue) {
+    mag_fld.x = std::numeric_limits<double>::quiet_NaN();
+    mag_fld.y = std::numeric_limits<double>::quiet_NaN();
+    mag_fld.z = std::numeric_limits<double>::quiet_NaN();
+  } else {
+    mag_fld.x = magnetic_field[0] * 1e-4 - params_.mag_bias_x;
+    mag_fld.y = magnetic_field[1] * 1e-4 - params_.mag_bias_y;
+    mag_fld.z = magnetic_field[2] * 1e-4 - params_.mag_bias_z;
+  }
+
+  return mag_fld;
+}
+
+geometry_msgs::msg::Vector3 PantherImuSensor::ParseGyration(const double angular_rate[3])
+{
+  geometry_msgs::msg::Vector3 ang_vel;
+
+  ang_vel.x = angular_rate[0] * (M_PI / 180.0);
+  ang_vel.y = angular_rate[1] * (M_PI / 180.0);
+  ang_vel.z = angular_rate[2] * (M_PI / 180.0);
+  return ang_vel;
+}
+
+geometry_msgs::msg::Vector3 PantherImuSensor::ParseAcceleration(const double acceleration[3])
+{
+  geometry_msgs::msg::Vector3 lin_acc;
+
+  lin_acc.x = -acceleration[0] * G;
+  lin_acc.y = -acceleration[1] * G;
+  lin_acc.z = -acceleration[2] * G;
+  return lin_acc;
+}
+
+void PantherImuSensor::HandleFirstDataCallback(
+  const geometry_msgs::msg::Vector3 & mag_compensated, const geometry_msgs::msg::Vector3 & lin_acc,
+  const double timestamp_s)
+{
+  if (
+    not std::isfinite(mag_compensated.x) or not std::isfinite(mag_compensated.y) or
+    not std::isfinite(mag_compensated.z)) {
+    // The nan values are provided when imu is not callibrated.
+    if (not imu_calibrated_) {
+      return;
+    } else {
+      throw std::runtime_error("Magnetometer has nan values.");
+    }
```

Comment:
```suggestion
  if (not imu_calibrated_) {
    return;
  }
  
  if (
    not std::isfinite(mag_compensated.x) or not std::isfinite(mag_compensated.y) or
    not std::isfinite(mag_compensated.z)) {
    throw std::runtime_error("Magnetometer has nan values.");
  }
```
Is it the same logic?

Reply:
yup